### ISSUE TITLE
Why doesn't IConfigurationProvider have an async Load method?

### ISSUE THREAD

**Issue by mariusGundersen** [dotnet/runtime#36018]
This might be a small or silly thing to ask about/for, but why isn't [`IConfigurationProvider.Load()`](https://github.com/aspnet/Extensions/blob/d656c4f7e22d1c0b84cab1b453c50ce73c89a071/src/Configuration/Config.Abstractions/src/IConfigurationProvider.cs#L38) async, returning a `Task`? Most configuration loading is going to be async, for example getting something from the filesystem or from the network. Now implementors are forced to use `GetAwaiter().GetResult()` in this method.

**Comment by maryamariyan** [comment]
The best next step on this would be to prepare the API proposal, usage, etc. in the issue page here (following https://github.com/dotnet/runtime/blob/master/docs/project/api-review-process.md).

**Comment by DkozhevnikovConfigit** [comment]
Hi all, any progress on that? 
We are working on our async application improvements and it would be super cool to have it async as well :)

**Comment by maryamariyan** [comment]
@DkozhevnikovConfigit do you have some usage examples in mind to help shape the API?

**Comment by DkozhevnikovConfigit** [comment]
@maryamariyan it's really simple example. 
We are storing our configuration within the DB, so when we want to load it in async manner we call GetAwaiter().GerResult() in Load and TryGet methods.

```
public class OurConfigurationProvider : ConfigurationProvider
{
    public override void Load()
    {
      Data = _db.GetConfigurationsAsync().GetAwaiter().GerResult();
    }

    public override bool TryGet(string key, out string value)
    {
      if (!Data.Contains(key))
      {
        var configuration = _db.GetConfigurationAsync(key).GetAwaiter().GerResult();
        if (configuration != null)
        {
          Data[key] = configuration;
        }
      }
        
      value = Data[key];
      return true;
    }
}
```
so it would be cool to do it async like:
```
public class OurConfigurationProvider : ConfigurationProvider
{
    public override async Task LoadAsync()
    {
      Data = await _db.GetConfigurationsAsync();
    }

    public override async ValueTask<bool> TryGetAsync(string key, out string value)
    {
      if (!Data.Contains(key))
      {
        var configuration = await _db.GetConfigurationAsync(key);
        if (configuration != null)
        {
          Data[key] = configuration;
        }
      }
        
      value = Data[key];
      return true;
    }
}
```

**Comment by eduardgrinberg** [comment]
Hi, any update on this?

We are missing this option as well. We're loading configuraition from http endpoint and would be really nice if we could use async here.

Thanks

**Comment by JasonBock** [comment]
FWIW I just ran into this with internal code at my employer. I also would like to see this have async APIs.

**Comment by KieranDevvs** [comment]
This really needs to be added. I'm loading a large configuration from a database, its not nice having to force other async API's to run synchronously because the BCL has gaps where it doesn't support it.

**Comment by piekstra** [comment]
I've just encountered this as well - we want to have our `appsettings.json` use paths to secrets in AWS Secrets Manager - with a custom `IConfigurationSource` and `ConfigurationProvider`, we are able to resolve those paths to actual secrets but should be able to do so asynchronously.

**Comment by dominikjeske** [comment]
I also vote for this.

**Comment by eventpromo** [comment]
Yes, i'm waiting for this feature as well

**Comment by niyazz** [comment]
I agree with the comments above and also need that feature to get a configuration with a custom provider over HTTP calls.

**Comment by f135ta** [comment]
At least this ticket has only been open for 3.5 yrs.......

**Comment by MCKanpolat** [comment]
I'm using Vault for getting secrets via http so async implementation is very useful such this scenarios.

**Comment by dazinator** [comment]
I've implemented an async adaptor over here: https://github.com/dazinator/Dazinator.Extensions.Configuration/blob/master/src/Dazinator.Extensions.Configuration.Async/AsyncConfigurationProviderAdaptor.cs#L19

Basically my adaptor supports asynchronous reloads - to do that I had to develop an async version of ChangeToken.OnChange.

However the initial load remains synchronous because it's called during `Build()` which is a synchronous call - the way I intend to solve this is that when you program starts prior to building the configuration you'd have to create the async provider and call something on it like InitialiseAsync so it could populate its initial values - that way when you call Build() and the initial sy synchronous Load() is called the provider could return its initial set of preloaded config and not an empty dictionary as I currently have it.

This should all not be necessary so +1

**Comment by stevo-knievo** [comment]
I have a similar scenario like @MCKanpolat. An `async` implementation would be really useful. Plus one for async!

**Comment by Edgaras91** [comment]
Does anyone know how I could safely fetch config asynchronously from `TryGet` override?
```
     public override bool TryGet(string key, out string value)
      {
          if (Data.ContainsKey(key))
          {
              //can't await? 
              FetchConfigAsync(key).RunSynchronously();
          }

          return base.TryGet(key, out value);
      }
```

**Comment by Rabadash8820** [comment]
@Edgaras91 You should probably open a separate Issue or ask your question on a Q&A site like StackOverflow. I'm not really sure what you're trying to do. `FetchConfigAsync` is already async, so there's no need to wrap it in another `async` call, and your additional `Data.Contains` validation isn't actually doing anything as written. All that said, if you just declare `TryGet` `async` then you can use `await`.

**Comment by KieranDevvs** [comment]
> @Edgaras91 You should probably open a separate Issue or ask your question on a Q&A site like StackOverflow. I'm not really sure what you're trying to do. `FetchConfigAsync` is already async, so there's no need to wrap it in another async call, and your additional `Data.ContainsKey` validation isn't actually doing anything as written. All that said, if you just declare `TryGet` `async` then you can use `await`.

You cant declare the overridden method as async because the return type would need to be `Task` or `Task<T>` which wouldn't match the virtual signature.

**Comment by Rabadash8820** [comment]
> You cant declare the overridden method as async because the return type would need to be Task or Task<T> which wouldn't match the virtual signature.

Good point. In that case, something like the following could work:

```cs
public override bool TryGet(string key, out string value)
{
    if (Data.ContainsKey(key))
    {
            value = FetchConfigAsync(key).Result;
            return true;
    }

    return base.TryGet(key, out value);
}
```

**Comment by Edgaras91** [comment]
The point is to not use Task.Result or any other approach to "call async in
sync method". Is there another workaround?

**Comment by Rabadash8820** [comment]
@Edgaras91 How would callers get the desired `value` then? You either have to synchronously wait on the `FetchConfigAsync` task or use some kind of callback (which is what async/await is essentially doing under the hood). A workaround would probably be to use aggregation instead of inheritance. Then, you wouldn't be constrained by the `TryGet` signature of your base class and could instead have one that returns `async Task<bool>`.

But again, we are de-railing this GitHub Issue. I encourage you to open a separate Issue or StackOverflow question or you can DM me. I will not be discussing it further here and I encourage others to do the same.

**Comment by klemmchr** [comment]
What's the status of the issue? Almost every configuration provider is async in some way, so this would benefit a lot of providers. Even file providers should take advantage of an async provider.

Synchronously waiting for a Task result might be fine in a lot of scenarios (like ASP.NET) but is not possible for Blazor apps. This is a huge drawback if you want to load any remote configuration. Loading remote configurations is extremely common in client side apps, it's almost a must have.

**Comment by 0xced** [comment]
There's already an API proposal for an async configuration provider: https://github.com/dotnet/runtime/issues/79193

Maybe we could try to vote and comment over there to give it some momentum. 🤞

**Comment by robertmclaws** [comment]
@davidfowl Any chance we could make this Async in .NET 10? Both Azure App Settings and AWS Parameter Store ConfigProviders would benefit immensely from this fix.

**Comment by aradalvand** [comment]
Yet another ping; configuration loading is almost invariably IO-based. Not having an async `Load` method makes no sense.

**Comment by Seabizkit** [comment]
ping

**Comment by lumimario** [comment]
ping
